{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "# Individual-level genotype-expression data generation\n",
    "\n",
    "## Goal\n",
    "This pipeline aims to perform cis-eQTL fine-mapping using individual-level genotype-expression data.\n",
    "\n",
    "## Overview of methods\n",
    "Specifically, we are to organize the genotype, expression, and covariates data into a gene-wise form, i.e., a data file per gene. Besides, we need the expression data 'perpendicular' to covariates, which will be used for regression with genotype, in instrument variable idea.\n",
    "\n",
    "## Details of methods\n",
    "covariates (`Z`): This is independent of genes, so this is the same among files and can be directly copied from the original covariates data.\n",
    "\n",
    "expression (`y`): As `grep` is too time-consuming, we construct `gene:line_number` as preprocessing. For each gene, the `line_number` is the line numbers of the gene in the #{tissues} expression data files. As a result, to extract the expression of each gene, it is only the job of reading the lines in the files, and combining the lines together.\n",
    "\n",
    "genotype (`X`): Obtain TSS of all genes as preprocessing. Once querying a gene, take the TSS of the gene, extracting SNPs within a predefined flanking region around the TSS and treating their genotypes as X.\n",
    "\n",
    "expression 'perpendicular' to covariates (`y_res`): The part of expression data that is independent from covariates.\n",
    "\n",
    "### Crystallized plan for generating y with the main step `gene:line_number`\n",
    "\n",
    "1) extract the column of ensembl_gene_id in the 49 expression files by `pandas`\n",
    "\n",
    "2) assign `gene:line_number` for each gene in each tissue based on the extracted column just now\n",
    "\n",
    "3) for each gene, extract the row of expression of the gene with sample names in each tissue using `sed`\n",
    "\n",
    "4) that's it!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "# Documentation\n",
    "\n",
    "## Input\n",
    "genotype data all in a file and expression and covariates data organized tissue-wise.\n",
    "\n",
    "- genotype: .vcf.gz\n",
    "\n",
    "    #CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  GTEX-1117F      GTEX-111CU      GTEX-111FC      GTEX-111VG      GTEX-111YS      GTEX-1122O\n",
    "    chr1    13526   chr1_13526_C_T_b38      C       T       .       PASS    AN=1676;AF=0.000596659;AC=1     GT      0|0     0|0     0|0     0|0     0|0     0|0\n",
    "    \n",
    "\n",
    "- expression: .bed.gz\n",
    "\n",
    "    #chr    start   end     gene_id GTEX-1117F      GTEX-111CU      GTEX-111FC      GTEX-111VG      GTEX-111YS      GTEX-1122O      GTEX-1128S\n",
    "    chr1    29552   29553   ENSG00000227232.5       1.3135329173730264      -0.9007944960568992     -0.29268956046586164    -0.7324113622418431     -0.27475411245874887    -0.6990255198601908     0.18188866123299216\n",
    "\n",
    "\n",
    "- covariates: .txt\n",
    "\n",
    "    ID      GTEX-1117F      GTEX-111CU      GTEX-111FC      GTEX-111VG      GTEX-111YS\n",
    "    PC1     -0.0867  0.0107  0.0099  0.0144  0.0154\n",
    "    PC2     -0.0132 -0.0026 -0.0050 -0.0081 -0.0093\n",
    "    \n",
    "\n",
    "## Output\n",
    "An RDS file containing genotype specific for each gene and expression and covariates of each gene.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "kernel": "bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: sos run Multi_tissues.ipynb [workflow_name | -t targets] [options] [workflow_options]\n",
      "  workflow_name:        Single or combined workflows defined in this script\n",
      "  targets:              One or more targets to generate\n",
      "  options:              Single-hyphen sos parameters (see \"sos run -h\" for details)\n",
      "  workflow_options:     Double-hyphen workflow-specific parameters\n",
      "\n",
      "Workflows:\n",
      "  preprocessing\n",
      "  extract\n",
      "\n",
      "Sections\n",
      "  preprocessing:        preprocessing: prepare gene expressions, gene TSS.\n",
      "    Workflow Options:\n",
      "      --cwd . (as path)\n",
      "      --expression-directory '/project/compbio/GTEx_dbGaP/GTEx_Analysis_2017-06-05_v8/eqtl/GTEx_Analysis_v8_eQTL_expression_matrices'\n",
      "      --genotype-file '/project/compbio/GTEx_dbGaP/GTEx_Analysis_2017-06-05_v8/genotypes/WGS/variant_calls/GTEx_Analysis_2017-06-05_v8_WholeGenomeSeq_838Indiv_Analysis_Freeze.SHAPEIT2_phased.vcf.gz'\n",
      "  extract:              extract Z, y, X from covariates, expression and genotype\n",
      "                        data and generate y_res\n",
      "    Workflow Options:\n",
      "      --ensembl-gene-id VAL (as str, required)\n",
      "      --cwd . (as path)\n",
      "      --expression-directory '/project/compbio/GTEx_dbGaP/GTEx_Analysis_2017-06-05_v8/eqtl/GTEx_Analysis_v8_eQTL_expression_matrices'\n",
      "      --genotype-file '/project/compbio/GTEx_dbGaP/GTEx_Analysis_2017-06-05_v8/genotypes/WGS/variant_calls/GTEx_Analysis_2017-06-05_v8_WholeGenomeSeq_838Indiv_Analysis_Freeze.SHAPEIT2_phased.vcf.gz'\n",
      "      --covariates-directory '/project/compbio/GTEx_dbGaP/GTEx_Analysis_2017-06-05_v8/eqtl/GTEx_Analysis_v8_eQTL_covariates'\n",
      "      --ntissues 49 (as int)\n"
     ]
    }
   ],
   "source": [
    "sos run Multi_tissues.ipynb -h"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Minimal working example\n",
    "**You can edit and change the following 2 bash variables.** First edit and run the following 2 bash variables.\n",
    "```\n",
    "work_dir=/scratch/midway2/chj1ar/GTEx_Analysis_v8_eQTL_expression_genewise\n",
    "ensembl_gene_id=ENSG00000140265.12\n",
    "```\n",
    "Then run as follows.\n",
    "```\n",
    "sos run Multi_tissues.ipynb preprocessing --cwd $work_dir\n",
    "sos run Multi_tissues.ipynb extract --cwd $work_dir --ensembl-gene-id $ensembl_gene_id\n",
    "```\n",
    "\n",
    "If you want to generate output files for all genes, run as follows. (**you can edit and change the bash variable $work_dir**)\n",
    "```\n",
    "work_dir=/scratch/midway2/chj1ar/GTEx_Analysis_v8_eQTL_expression_genewise\n",
    "sos run Multi_tissues.ipynb preprocessing --cwd $work_dir\n",
    "sos run Multi_tissues.ipynb parallelize --cwd $work_dir --no-self-defined\n",
    "```\n",
    "\n",
    "If you want to generate output files for a list of self-defined genes, collect the list of genes into `ensembl_gene_id_run.txt` in work_dir with each gene ID a line, possibly after preprocessing, and run as follows. (**you can edit and change the bash variable $work_dir**)\n",
    "```\n",
    "work_dir=/scratch/midway2/chj1ar/GTEx_Analysis_v8_eQTL_expression_genewise\n",
    "sos run Multi_tissues.ipynb preprocessing --cwd $work_dir\n",
    "sos run Multi_tissues.ipynb parallelize --cwd $work_dir --self-defined\n",
    "```\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# preprocessing: prepare gene expressions, gene TSS.\n",
    "[preprocessing]\n",
    "depends: R_library(\"biomaRt\"), R_library(\"data.table\")\n",
    "parameter: cwd = path()\n",
    "parameter: expression_directory = '/project2/compbio/GTEx_dbGaP/GTEx_Analysis_2017-06-05_v8/eqtl/GTEx_Analysis_v8_eQTL_expression_matrices'\n",
    "parameter: genotype_file = '/project2/compbio/GTEx_dbGaP/GTEx_Analysis_2017-06-05_v8/genotypes/WGS/variant_calls/GTEx_Analysis_2017-06-05_v8_WholeGenomeSeq_838Indiv_Analysis_Freeze.SHAPEIT2_phased.vcf.gz'\n",
    "input: expression_directory, genotype_file\n",
    "output: 'ensembl_gene_id.txt', 'samplenames_X.txt', 'gene-line_number/*.csv', 'gene_TSS.csv'\n",
    "bash: expand = '${ }', workdir = cwd\n",
    "    # `ensembl_gene_id.txt`\n",
    "    for file in ${expression_directory}/*.gz; do zless \"$file\" | grep -oh \"ENSG\\w*\\.\\w*\" >> ensembl_gene_id_original.txt ; done\n",
    "    sort -u ensembl_gene_id_original.txt > ensembl_gene_id.txt\n",
    "    rm ensembl_gene_id_original.txt\n",
    "    \n",
    "    # mkdir gene-line_number (in cwd)\n",
    "    mkdir gene-line_number\n",
    "    \n",
    "    # `samplenames_X.txt`\n",
    "    zless ${genotype_file} | sed '3386q;d' > samplenames_X.txt\n",
    "\n",
    "python3: expand = '${ }', workdir = cwd\n",
    "    # `gene:line_number`\n",
    "    \n",
    "    with open(\"ensembl_gene_id.txt\") as f:\n",
    "        ensembl_gene_id = [line.rstrip(\"\\n\") for line in f]\n",
    "    \n",
    "    import glob\n",
    "    import pandas as pd\n",
    "    filenames_y = glob.glob(\"${expression_directory}/*.gz\")\n",
    "    filenames_y.sort()\n",
    "    gene_linenumber = {} # Now gene_linenumber is a list of 49 dict, each of which has 20000 items. But if gene_linenumber is a dict of 20000 dict, each of which has 49 items, then it can be saved into 20000 files!\n",
    "    fields = ['gene_id', ]\n",
    "    for j in range(len(ensembl_gene_id)): # construct ~20000 dict, each of which has 49 items\n",
    "        gene_linenumber[ensembl_gene_id[j]] = {}\n",
    "\n",
    "    for i in range(len(filenames_y)): # construct each of the 49 `gene:line_number`\n",
    "        gene_id = pd.read_csv(filenames_y[i], sep='\\t', skipinitialspace=True, usecols=fields)\n",
    "        for j in range(gene_id.shape[0]): # go over lines\n",
    "            gene_linenumber[gene_id['gene_id'].tolist()[j]][i] = j + 2\n",
    "            # print('%s at line %d in file %d' % (gene_id['gene_id'].tolist()[j], j + 2, i))\n",
    "    \n",
    "    import csv\n",
    "    for j in range(len(gene_linenumber)): # write ~20000 files\n",
    "        with open(('gene-line_number/%s.csv' % ensembl_gene_id[j]), 'w') as f:\n",
    "            writer = csv.writer(f)\n",
    "            for key, value in gene_linenumber[ensembl_gene_id[j]].items():\n",
    "                writer.writerow([key, value])\n",
    "        \n",
    "R: expand = True, workdir = cwd\n",
    "    # `gene_TSS.csv`\n",
    "    ensembl_gene_id <- data.table::fread(file = \"ensembl_gene_id.txt\", sep = \"\\n\", quote = \"\", header = FALSE)\n",
    "    mart <- biomaRt::useDataset(\"hsapiens_gene_ensembl\", biomaRt::useMart(\"ensembl\"))\n",
    "    gene_TSS <- biomaRt::getBM(attributes = c(\"chromosome_name\", \"transcript_start\", \"ensembl_gene_id\", \"ensembl_gene_id_version\"), filters = \"ensembl_gene_id_version\", values = ensembl_gene_id, mart = mart)\n",
    "    rm(ensembl_gene_id)\n",
    "    rm(mart)\n",
    "    write.table(x = gene_TSS, file = \"gene_TSS.csv\", sep = '\\t', quote = FALSE, col.names = TRUE, row.names = FALSE)\n",
    "    rm(gene_TSS)\n",
    "    # gene_TSS_retrieval <- read.table(file = \"gene_TSS.csv\", sep = '\\t', quote = \"\", stringsAsFactors = FALSE, header = TRUE)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# parallelize genes\n",
    "[parallelize]\n",
    "parameter: cwd = path()\n",
    "parameter: self_defined = bool\n",
    "import os\n",
    "if self_defined:\n",
    "    with open(os.path.join(cwd, 'ensembl_gene_id_run.txt')) as f:\n",
    "        ensembl_gene_id = f.readlines()\n",
    "    \n",
    "else:\n",
    "    with open(os.path.join(cwd, 'ensembl_gene_id.txt')) as f:\n",
    "        ensembl_gene_id = f.readlines()\n",
    "    \n",
    "ensembl_gene_id_list = [x.strip('\\n') for x in ensembl_gene_id]\n",
    "input: for_each={'x': ensembl_gene_id_list}\n",
    "sos_run('extract', ensembl_gene_id=ensembl_gene_id_list[_index])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# extract Z, y, X from covariates, expression and genotype data and generate y_res\n",
    "[extract]\n",
    "depends: R_library(\"data.table\")\n",
    "parameter: cwd = path()\n",
    "parameter: ensembl_gene_id = str\n",
    "parameter: expression_directory = '/project2/compbio/GTEx_dbGaP/GTEx_Analysis_2017-06-05_v8/eqtl/GTEx_Analysis_v8_eQTL_expression_matrices'\n",
    "parameter: genotype_file = '/project2/compbio/GTEx_dbGaP/GTEx_Analysis_2017-06-05_v8/genotypes/WGS/variant_calls/GTEx_Analysis_2017-06-05_v8_WholeGenomeSeq_838Indiv_Analysis_Freeze.SHAPEIT2_phased.vcf.gz'\n",
    "parameter: covariates_directory = '/project2/compbio/GTEx_dbGaP/GTEx_Analysis_2017-06-05_v8/eqtl/GTEx_Analysis_v8_eQTL_covariates'\n",
    "parameter: ntissues = 49\n",
    "input: expression_directory, genotype_file, covariates_directory\n",
    "# output: \"Multi_Tissues.%s.RDS\" % ensembl_gene_id\n",
    "bash: expand = '${ }', workdir = cwd\n",
    "    # X\n",
    "    awk '$4 ~ /${ensembl_gene_id}/ {print $0}' gene_TSS.csv | head -n 1 | awk '{$3=$2+1000000} {$2=$2-1000000} {print \"chr\"$1\":\"$2\"-\"$3}' > ${ensembl_gene_id}_TSSregion.txt\n",
    "    tabix ${genotype_file} \"$(< ${ensembl_gene_id}_TSSregion.txt)\" | sed -e 's/0|0/0/g' -e 's/0|1/1/g' -e 's/1|0/1/g' -e 's/1|1/2/g' > ${ensembl_gene_id}_genotype.txt # still remain a potential problem of multi-allelic case.\n",
    "    \n",
    "R: expand = True, workdir = cwd\n",
    "    # Z\n",
    "    filenames <- list.files(path = \"{covariates_directory}\", pattern = \"*.txt\", full.names = TRUE)\n",
    "    Z <- lapply(filenames, function(x) t(as.matrix(read.table(file = x, header = TRUE, sep = '\\t', quote = \"\", row.names = 1))))\n",
    "    for (i in 1:{ntissues}) {{ # the format of sample names Z was changed somehow, from \"GTEX-*\" to \"GTEX.*\", so we need to convert it back.\n",
    "        rownames(Z[[i]]) <- gsub(pattern = \"GTEX.\", replacement = \"GTEX-\", x = rownames(Z[[i]]))\n",
    "    }}\n",
    "    \n",
    "    # y\n",
    "    \n",
    "    filenames_y <- list.files(path = \"{expression_directory}\", pattern = \"*.gz$\", full.names = TRUE)\n",
    "    line_numbers <- read.table(file = \"gene-line_number/{ensembl_gene_id}.csv\", sep = ',', header = FALSE, row.names = 1, quote = \"\")\n",
    "    extract_y <- function(i) {{\n",
    "        yi <- data.table::fread(file = filenames_y[i], skip = line_numbers[i,] - 1, nrows = 1)\n",
    "        samplenames_yi <- data.table::fread(file = filenames_y[i], skip = 0, nrows = 1)\n",
    "        colnames(yi) <- colnames(samplenames_yi)\n",
    "        yi <- t(as.matrix(yi))\n",
    "        yi <- yi[-1:-4, , drop = FALSE]\n",
    "        return(yi)\n",
    "    }}\n",
    "    \n",
    "    y <- lapply(1:{ntissues}, extract_y)\n",
    "    \n",
    "    ## sample names matching between y and Z\n",
    "    samplenamesmatching <- function(x, reference) {{\n",
    "        lapply(1:{ntissues}, function(i) x[[i]][match(rownames(reference[[i]]), rownames(x[[i]])), , drop = FALSE])\n",
    "    }}\n",
    "    \n",
    "    y_matchZ <- samplenamesmatching(x = y, reference = Z)\n",
    "    y <- y_matchZ\n",
    "    \n",
    "    # X\n",
    "    X <- read.csv(file = \"{ensembl_gene_id}_genotype.txt\", sep = '\\t', header = FALSE, row.names = 3, stringsAsFactors = FALSE)\n",
    "    X <- X[,-(1:8)]\n",
    "    samplenames_X <- scan(file = \"samplenames_X.txt\", what = character(), quote = \"\")\n",
    "    samplenames_X <- samplenames_X[-(1:9)]\n",
    "    colnames(X) <- samplenames_X\n",
    "    X <- as.matrix(x = X)\n",
    "    X <- t(X)\n",
    "    \n",
    "    # y_res\n",
    "    y_res <- lapply(1:{ntissues}, function(i) .lm.fit(x = Z[[i]], y = y[[i]])$residuals)\n",
    "\n",
    "    # save\n",
    "    saveRDS(object = list(X = X, y = y, Z = Z, y_res = y_res), file = \"Multi_Tissues.{ensembl_gene_id}.RDS\")\n",
    "\n",
    "bash: expand = '${ }', workdir = cwd\n",
    "    # remove intermediate files\n",
    "    rm ${ensembl_gene_id}_TSSregion.txt\n",
    "    rm ${ensembl_gene_id}_genotype.txt\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "codemirror_mode": "sos",
   "file_extension": ".sos",
   "mimetype": "text/x-sos",
   "name": "sos",
   "nbconvert_exporter": "sos_notebook.converter.SoS_Exporter",
   "pygments_lexer": "sos"
  },
  "sos": {
   "kernels": [
    [
     "SoS",
     "sos",
     "",
     "",
     "sos"
    ],
    [
     "bash",
     "bash",
     "bash",
     "",
     "shell"
    ]
   ],
   "panel": {
    "displayed": true,
    "height": 0
   },
   "version": "0.19.13"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
